share/doc/wake/logging.md

```diff
@@ -0,0 +1,111 @@
+It also controls which *actual* stream it is going to coming out of the executing `wake` process
+(stderr or stdout, or other file descriptors 3-5).
+
+`wake` has a number of built in loggers in wake which all start with `log...`.
+You can also make your own logger using `mkLogLevel`.
+You can use this to customize the color you want it to be shown as (if at all),
+and giving more fine-grained control over which wake output stream it should go to.
+
+For stdout, the following table should be read top to bottom considering the command line arguments you've passed to wake.
+If you match a line in the table, stop!
+This is what will appear on the console as you watch wake execute, and what is sent to wake's stdout:
+
+|Flag               | logDebug | logInfo | logEcho | logReport | logWarning | logError | logNever | mkLogLevel "foo" Green|
+|-------------------|----------|---------|---------|-----------|------------|----------|----------|----------------------
+|Color on Console   |  Blue    |  Dim    |  default| Magenta   |  Yellow    |   Red    |          |      Green      |          
+|--stdout="foo"     |          |         |         |           |            |          |          |         x       |
+|--stdout="foo,info" |         |    x    |         |           |            |          |          |         x       |
+|--debug            |      x   |    x    |    x    |     x     |      x     |    x     |          |                 |
+|--verbose          |          |    x    |    x    |     x     |      x     |    x     |          |                 |
+|--quiet            |          |         |         |           |            |    x     |          |                 |
+|--no-tty<sup>*</sup>|         |         |    x    |     x     |      x     |    x     |          |                 |
+|(default)          |          |         |         |     x     |      x     |    x     |          |                 |
+
+<sup>*</sup> `--no-tty` also prevents any coloring from being applied.
+
+
+Unless overridden with `--stderr=...` the *only* thing that ever appears on stderr coming out of wake is `logError`,
+and it always appears there.
+
+The additional command line arguments `--fd:3`, `--fd:4`, `--fd:5` can be used to output to file descriptors 3, 4, 5.
+One example shell command would be, for some wake code that had `mkLogLevel "foo"`:
+
+```
+wake --fd3="foo" mywakecode 3&>foo.txt
+```
+
+These can be concatenated with commas and include the normal log levels, so to capture both `mkLogLevel "foo"` and `logInfo` to a file:
+
+```
+wake --fd3="foo,info" 3&>foo.txt
+```
+
+## What Gets Sent To Loggers
+
+### Job Execution
+
+Every `Job` run by wake tracks its own stdout and stderr,
+and when the underlying process prints to stdout or stderr,
+it will go to the `Job`-specific stdout and stderr,
+not the stdout and stderr of the `wake` process itself.
+Every `Job` also has an "echo", which is the command that the job is executing.
```

Comment:
I don't think I actually understand what this last part means. Is this some wake feature that I don't know about? What exactly is the "echo"? Is it a property on the `Job` tuple? Is it a column in the jobs table of the job database? What type of thing is an "echo"? stdout and stderr are file-like streams on Unix processes that most programmers are very familiar with. "echo" clearly is not one of those, but you don't identify the type thing that "echo" is, so this sentence is difficult to understand.

The text in the following paragraphs don't really help me much either, so I think something needs to be clarified, since seeing "echo" mentioned alongside "stdout" and "stderr" is very unexpected to me.

Reply:
I learned this too in the process of writing the doc, The 'echo' is referring to/identical to the command you're actually going to execute with the job (if I understand correctly). I could just call it "the command", but setPlanEcho is what you use to set where it is logged, so 🤷 . In other words, in the database you are storing a bunch of things, like the environment variables, the command you are going to execute, the stderr and stdout. The "echo" is identical to "the command you are going to execute" (I think? maybe there is other stuff in the echo like the label of the job or something?)

Reply:
https://github.com/sifive/wake/blob/63151ddf4a75ed597a9b92588569be6fe28fa677/share/wake/lib/system/sources.wake#L45

Reply:
But what "type" of thing is an "echo"? There's no equivalent concept in Unix, so the explanation above this sentence that talks about the "underlying process" doesn't make sense when applied to "echo".

It sounds like this thing is basically just equivalent to a wake `printLogLevel` call that wake does just before running the Job with the command that the Job is about to run. If this is the case, then I think it would be best for you to explain it separately from stdout and stderr, because those actually correspond to things in the Unix process model (streams that are accessible via file descriptors within the process), whereas "echo" appears to only exist in wake-land.

Reply:
> It sounds like this thing is basically just equivalent to a wake printLogLevel call that wake does just before running the Job with the command that the Job is about to run.

Sort of, but it is controlled in analagous way as SetPlanStderr and SetPlanStdout. I could break this out separately into the "other stuff" section (i had it there originally, but then moved it back here, because this DOES appear in the database as it goes with an executing job). And, you set it like you set setPlanStderr, with setPlanEcho. 

Maybe I can just not introduce a noun of 'echo' and just call it 'the command you are going to execute' except for saying you override that with `setPlanEcho`.

Reply:
or "command echo" or "echoed command"?

Reply:
Reworded all this, I think it is more clear now (I brought the 'echo' part back down to the "other stuff" area)